packages/next-swc/crates/core/src/relay.rs

```diff
@@ -0,0 +1,157 @@
+        }
+    }
+}
+
+struct Relay {
+    config: Config,
+    file_name: FileName,
+}
+
+fn pull_first_operation_name_from_tpl(tpl: TaggedTpl) -> Option<String> {
+    tpl.tpl
+        .quasis
+        .into_iter()
+        .filter_map(|quasis| {
+            let template_string_content = String::from(quasis.raw.value.to_string());
+            let split_content = template_string_content.split(" ").collect::<Vec<&str>>();
+
+            let operation = split_content
+                .chunks(2)
+                .filter_map(|slice| {
+                    if slice.len() == 1 {
+                        return None;
+                    }
+
+                    let word = slice[0];
+                    let next_word = slice[1];
+
+                    if word == "query" || word == "subscription" || word == "mutation" {
+                        return Some(String::from(next_word));
+                    }
+
+                    None
+                })
+                .next();
+
+            return operation;
+        })
+        .next()
+}
+
+fn build_require_expr_from_path(path: String) -> Expr {
+    Expr::Call(CallExpr {
+        span: Default::default(),
+        callee: ExprOrSuper::Expr(Box::new(Expr::Ident(Ident::new(
+            js_word!("require"),
+            Span::default(),
+        )))),
+        args: vec![ExprOrSpread {
+            spread: None,
+            expr: Box::new(Expr::Lit(Lit::Str(Str {
+                span: Default::default(),
+                value: JsWord::from(path),
+                has_escape: false,
+                kind: Default::default(),
+            }))),
+        }],
+        type_args: None,
+    })
+}
+
+impl Fold for Relay {
+    fn fold_expr(&mut self, expr: Expr) -> Expr {
+        let expr = expr.fold_children_with(self);
+
+        match expr.clone() {
+            Expr::TaggedTpl(tpl) => {
+                if let Some(built_expr) = self.build_call_expr_from_tpl(tpl) {
+                    built_expr
+                } else {
+                    expr
+                }
+            }
+            _ => expr,
+        }
+    }
+}
+
+impl Relay {
+    fn build_call_expr_from_tpl(&mut self, tpl: TaggedTpl) -> Option<Expr> {
+        if let Expr::Ident(ident) = *tpl.tag.clone() {
+            if ident.sym.to_string() != "graphql" {
+                return None;
+            }
+        }
+
+        let operation_name = pull_first_operation_name_from_tpl(tpl);
+
+        if let (Some(operation_name), Real(source_path_buf)) =
+            (operation_name, self.file_name.clone())
+        {
+            let path_to_source_dir = source_path_buf.parent().unwrap();
+            let generated_file_name = format!(
+                "{}.graphql.{}",
+                operation_name,
+                self.config.file_extension().clone()
+            );
+
+            let fully_qualified_require_path = match self.config.artifact_directory.clone() {
+                Some(artifact_directory) => std::env::current_dir()
```

Comment:
is there a way to get the root of the application without calling `std::env::current_dir()` from some configs/env variables?

Reply:
I suppose `relay.json` should always be at the root of the project and I could read from there.

Reply:
I have a big monorepo, with many Next.js app's all sharing a single artifacts directory. Will that continue to work?

```
apps/
├─ frontend1/
│  ├─ next.config.js
├─ frontend2/
│  ├─ next.config.js
package.json
relay.config.json
artifactsDirectory/
```

We use a combination of tsconfig.js#paths in each frontend application pointing `artifactsDirectory: ../../artifactsDirectory/$1` ... and `next.config.js#experimental.externalDir` allowing us to import typescript from _outside_ of our application's rootDir.

Sharing types across frontends is very handy, as we have a design system with many fragment components (data colocation ftw) that can be used across different frontend applications.

Reply:
@alunyov I think we could support this if we try to support `MultiProjectConfig`. I think the problem right now is that I don't have a way to determine which project to use for a given source file. Is there something you guys expose for this?

Reply:
Each of our Next applications have a `babelrc` that looks like this:

```
  plugins: [
    [
      'relay',
      {
        src: './',
        artifactDirectory: '../../artifactsDirectory',
        extensions: ['js', 'ts', 'tsx'],
        include: ['**/*'],
        schema: '../apps/backend/schemas/__generated__/main.graphql',
        watchman: false,
        language: 'typescript',
      },
    ],
  ],
```

We then have a custom watcher that runs @ the root directory watches code in the different applications and manually re invoke `relay-compiler`.

What's `MultiProjectConfig`?

Reply:
Hmmm, seems like I can use something from here to determine which project config to use for a given file https://sourcegraph.com/github.com/facebook/relay/-/blob/compiler/crates/relay-compiler/src/file_source/file_categorizer.rs?L412

Reply:
@hanford If you could setup a minimal repro, that would be super helpful to guide my integration tests for multi project configs.

Reply:
@hanford I was able to create an integration test from what you've described. Check out the multi project integration test and let me know if that is more or less what your team has

Reply:
@tbezman This looks nearly identical to what we have though my applications aren't using a MultiProject config, (maybe we should?).

I've created a repo that's really close to what we have: https://github.com/hanford/many-relay-next

I suspect that as long as my frontend applications can reference an artifacts directory from outside of their individual source codes we'll be working

--

If you have a compiled SWC + Relay binary, I'd also be down to try and give it a spin in my private repo and can report back

Reply:
@hanford It's more than the binary you'll need. I would recommend pulling this branch down and using `yarn link` to set it up. I'm not sure which packages you'll need to build + link.

This is what I do to build the binary.
`cd packages/next-swc; yarn build-native`

Reply:
Tested, my workflow seems to work pending some fixes I've mentioned below. 

SO EXCITED!